ivadomed/ivadomed
# New loader: Refactor BidsDataset classes

### Context

Fixes #596 
This PR is part of the loader refactoring project to include microscopy data (#304) and to migrate to HDF5 loader (#474).

It aims to refactor `BidsDataset`, `Bids3DDataset` and `BIDStoHDF5` classes to use the new loader pipeline (i.e. BidsDataframe + new splitting methods) to create `filename_pairs`.
The resulting `filename_pairs` are the same as the `filename_pairs` in the current loader.

### Changes done in this PR:

As discussed previously, the changes are made in "new" files until they are integrated in the pipeline.
- Refactoring of `BidsDasaset` and `Bids3DDataset` in `loader/loader_new.py` file
- Refactoring of `load_bounding_box` and `generate_bounding_box_file` in `object_detection/utils_new.py` file.
- Refactoring of `BIDStoHDF5` and `HDF5Dataset` in `loader/adaptative_new.py`.
- Remove `ivadomed_id` column in `BidsDataframe`
- Split functions output lists of `filenames`  instead of list of `ivadomed_id`
- `BidsDataset` and `BIDStoHDF5` input lists of `filenames` instead of `participants_id` (I made it retrocompatible with partitcipant_id, old joblib files should work)
- All the other functions/classes in "_new" files are identical to the originals, except for some imports.

### To test the new pipeline:
- Since, this is not integrated with the current pipeline, you need to run the `dev/df_new_loader.py` script to load a validation dataset (works only for nifti files)
- The script uses `config_new_loader.json` config file. To use older config files, make sure to add the following parameters:
   - `"extensions": [".nii.gz"]` in loader_parameters
   -   `"split_method" : "participant_id"`, and `"data_testing": {"data_type": null, "data_value":[]}`, in split_dataset parameters
- I tested the `U-net 2D` with and without `roi_suffix`, `multichannel` and `bounding_box`, `FilmedUnet` with `mri_param metadata_choice` and `custom metadata_choice`, as well as `Unet-3D` and `Hemis`. It seems to work fine but it would be great to test with real experiment files in case I missed something.

### Next step:
- Integrate the changes made for the new loader in PR #538, #591, #648 and #687 to the current pipeline.
